### ISSUE TITLE
[BrainStorming as BS] 코드카타 아이디어 모집

### ISSUE THREAD

**Issue by odol87** [odol87/codeKaTa#1]
### 어떤 걸 해야할지 커멘트!!

**Comment by odol87** [comment]
- 소팅 구현
- 재귀 마스터
- 함수형 프로그래밍 ex) currying 을 활용한 함수 합성
- react 정복 -> react 문서 내용 이해한바 정리 (특히 고급 부분)

**Comment by odol87** [comment]
- 비동기로 들어오는 데이터 처리 다양한 방식으로 해보기
ex)
```
const menuIds = ['1', '2', '3']
const products: Map<string, products> = new Map()

for (menuId of menuIds) {
 const fetchedProducts = await fetchData(menuId)
 products.set(menuId, fetchedProducts)
}

const fetchData = async (menuId: string): Promise<product[]> => {
...
}
```

이 코드의 단점은? cc/ @euntak

**Comment by odol87** [comment]
[코드카타 아이템으로 TODO app 을 선택하지 않은 이유](https://woowabros.github.io/experience/2020/04/14/stop-making-todo-list.html)

**Comment by odol87** [comment]
- rxjs 붙여보기

**Comment by euntak** [comment]
> * 비동기로 들어오는 데이터 처리 다양한 방식으로 해보기
>   ex)
> 
> ```
> const menuIds = ['1', '2', '3']
> const products: Map<string, products> = new Map()
> 
> for (menuId of menuIds) {
>  const fetchedProducts = await fetchData(menuId)
>  products.set(menuId, fetchedProducts)
> }
> 
> const fetchData = async (menuId: string): Promise<product[]> => {
> ...
> }
> ```
> 
> 이 코드의 단점은? cc/ @euntak

이렇게 댓글을 달면 되나.. 일단 예상하는 결과가 다를 것 같은데,
for .. of 는 동기로 동작하고, 비동기로 동작하는 `fetchData`는 이벤트 큐로 들어가게 될테니 set에 들어가는 값 `fetchedProducts`는 `undefined`가 될 것 같네요.

그리구.. fetchData가 const로 정의 되어있기 때문에 호이스팅이 안되서 호출 자체가 안될 것 같네요

음, `fetchData` 의 함수 내부 구현에 따라서 다를 순 있겠지만 함수의 이름만 봤을 땐 서버로 부터 데이터를 받는다고 생각이 드는데 무조건 응답이 온다고 정해 놓은 것 같아서, 응답이 오지 않았을때에 대한 처리도 같이 해줘야 할 것 같습니당.

**Comment by euntak** [comment]
그리구.. fetchData가 const로 정의 되어있기 때문에 호이스팅이 안되서 호출 자체가 안될 것 같네요

**Comment by euntak** [comment]
아니 근데 테스트 해보니까.. for ... of 에서 비동기 처리가 되네요..? 근데 순차적으로 진행되는 비동기라, 중간에 오래 걸리는 비동기 처리가 있을 경우에는 병목이 일어날 것 같네요.

**Comment by euntak** [comment]
비동기 결과가 순서에 영향을 받지 않고, 병목이 일어나지 않도록 하는 해결 방법으로 `Promise.all` 사용

```js
const menuIds = ['31', '2', '1']
const products = new Map()

const promises = Promise.all(menuIds.forEach(async menuId => await fetchProduct(menuId)));

async function fetchProduct(menuId) {
    const fetchedProduct = await fetchData(menuId);
    products.set(menuId, fetchedProduct);
}

async function fetchData (menuId) {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(menuId);
        }, 100 * Number(menuId));
    });
}
```
